### DISCUSSION TITLE
Control with Alexa

### DISCUSSION THREAD

**Discussion by Krocko** [ekutner/home-connect-hass#124] [Ideas]
The input.select for the programs and the button entities are really nice and clean. This is working well inside Home Assistant.
But there is no chance to control the device with Alexa.
Could you also add switches for on/off and the programs? So we can use the input.select and button with Home Assistant and can use the switches to control the device with Alexa.

**Comment by ekutner** [comment]
The integration automatically creates the type of control which is appropriate for the value provided by the API. Binary values become switches, numeric values become numbers and textual values become selects. This is done automatically, the thing is that some values you would expect to be binary, like the power state of the appliance, are actually provided by the API as text so they become selects (because in that example there is also a "standby" state, not only on/off). 
Once the main functionality is verified to be working properly I'll look into adding Alexa support.

**Comment by Krocko** [comment]
Thank you very much. 
I am available for testing with a
Siemens CoffeeMaker TI957FX1DE
and a
Siemens WasherDryer WD14U592

**Comment by NODeeJay** [comment]
>  [...] Could you also add switches for on/off and the programs? So we can use the input.select and button with Home Assistant and can use the switches to control the device with Alexa.

When you just need it to start the command with voice you can always add a template (binary) sensor or template switch in HA. I use it that way to control many switches that are not switches in reality with Genie. This way you can also assign a more practical name to the "skill" and don't need to alter the original entity.
It adds a layer of complexity but helps tremendously when you e.g. change the appliance from Bosch to AEG or whatever because your skills will stay the same.

**Comment by ekutner** [comment]
Thanks @NODeeJay. It would be great if you can give a detailed example of how to do it for those who wish to integrate with Alexa.

**Comment by thomasadelhardt** [comment]
Thanks @NODeeJay for the suggestion with the template.
It might not be perfect, but I gave it a shoot.
A switch is now exposed to my voice assistant (Google Home), so I can start/stop my Siemens iQ500 dishwasher with voice commands.
Her is what I put into my configuration.yaml:

`
switch:
  - platform: template
    switches:
      iq500:
        friendly_name: "Opvaskemaskine"
        unique_id: "siemens_iq500_startstop"
        value_template: "{{ not is_state('button.012050388751000127_stop', 'unavailable') and is_state('button.012050388751000127_start_pause', 'unavailable') }}"
        availability_template: "{{ is_state('binary_sensor.012050388751000127_bsh_common_status_doorstate', 'off') and is_state('binary_sensor.012050388751000127_connected', 'on') and is_state('binary_sensor.012050388751000127_bsh_common_status_remotecontrolstartallowed', 'on') and is_state('binary_sensor.012050388751000127_bsh_common_status_remotecontrolactive', 'on') and not is_state('sensor.012050388751000127_bsh_common_status_operationstate', 'aborting') }}"
        icon_template: "mdi:dishwasher"
        turn_on:
          - service: switch.turn_on
            target:
              entity_id: switch.012050388751000127_bsh_common_setting_powerstate
          - service: button.press
            target:
              entity_id: button.012050388751000127_start_pause
        turn_off:
          service: button.press
          entity_id: button.012050388751000127_stop
`

Hopefully you can also make it work @ekutner. 
(Added screenshot because all the indents are lost)
![Screenshot from 2022-07-27 23-32-07](https://user-images.githubusercontent.com/70981717/181376738-e5aa738b-a9da-466e-af45-5bae5c47e941.png)

**Comment by cibernox** [comment]
I was playing with this and in node red I was able to create a thermostat-like appliance that supports on/off and to set a temperature. It is not perfect and I can’t change modes or set alarms, but 95% of the time I use heat from top and bottom so it’s not bad. And you can ask Alexa to turn it off in a certain amount of time, which fires an Alexa timer instead of a time on the oven, but that is also aceptable. 

However the node-red package I’m using is unmaintainded and this feels like something that should be baked in. 

What would require to expose ovens like thermostats natively to Alexa? That’s for now the best solution I’ve found other than creating a custom skill that requires you to use the obnoxious “Alexa, ask Home Assistant to turn on the oven at 220 degrees”. I really hate having to use the skill invocation name for this.

**Comment by ekutner** [comment]
Looks like this should be a discussion, not an issue, so moving it there.

**Comment by cibernox** [comment]
FWIW, I managed to get this working using `node-red-contrib-alexa-home-skill`. 

This node-red package (paired with the service in https://alexa-node-red.bm.hardill.me.uk/) allows you to create virtual alexa appliances with the capabilities you select. I created a thermostat that can be turned on and off and allows you to set a target temperature on it:
![Screenshot 2022-10-22 at 17 03 50](https://user-images.githubusercontent.com/265339/197346426-f2d0cdf9-bb78-481e-85d3-43f355427197.jpg)

Then you can create node-red flows that start with commands:
![Screenshot 2022-10-22 at 17 09 05](https://user-images.githubusercontent.com/265339/197346692-9313926f-7169-40cf-ac5c-0bed009015a8.jpg)


In this flow, first I have a security check that the home is occupied. This is me wearing a tinfloil hat but I don't ever want my oven to respond to voice commands unless me or some other adult is in the house.
Then I check what kind of command I've received (turn on / turn off / set temperature). When it's a simple turn on, I set the temperature to 180°C which I've decided it's a good default temperature. When it's a turn off I just it turn off. Then I get the temperature (set by the user or the default 180) and I check if the oven is already working, to know wether I should send `program_start` call or a `set_program_option` call.

This allows Alexa understand commands like:
- Alexa, turn on the oven.
- Alexa, turn off the oven.
- Alexa, turn on the oven at 220 degrees
- Alexa, set the oven to 220 degrees
- Alexa, turn off the oven un 20 minutes (note, this creates a timer within alexa, not a timer in the oven itself, which is not ideal)
and a few similar ones.

I went for this approach because I hate with passion having to say the skill name every time I have to interact with an appliance. In case the invocation of my custom skill is "Home assistant":
- Alexa, ask home assistant to turn on the oven.
- Alexa, ask home assistant to set the oven to 220 degrees.

It has many caveats that for me are worth having:
- You can't change the oven's program. I've decided that since heat from top and bottom is what I use 95% of the time, that's the program this integration will use.
- Since all programs need to be started with a time, I set this to 2h. This is long enough to roast almost anything I want but not an amount that if I forget to turn it off would ruin me.
- You can't turn on and off the fast preheat. In my case I decided I want it on all the time because I rarely, if ever, want my oven to be slower on purpose.
- Timers and alarms live within alexa, not inside the oven, so there won't be a bell ringing when the program ends, and in the unlikely scenario that alexa has a blip the oven would not turn off. So far that never happened, but all scenarios have to be considered.

In case this integration manages to do this natively without home assistant packages, until Alexa supports ovens natively, in case that ever happens, our only solution is to create a thermostat similar to the one I've described. Perhaps it should be the user's choice and not a default.
